core/services/request/files/FilesDataHandler.php

```diff
@@ -0,0 +1,235 @@
+<?php
+
+namespace EventEspresso\core\services\request\files;
+
+use EventEspresso\core\services\collections\CollectionDetails;
+use EventEspresso\core\services\collections\CollectionDetailsException;
+use EventEspresso\core\services\collections\CollectionInterface;
+use EventEspresso\core\services\collections\CollectionLoader;
+use EventEspresso\core\services\collections\CollectionLoaderException;
+use EventEspresso\core\services\request\Request;
+use UnexpectedValueException;
+
+/**
+ * Class FilesDataHandler
+ *
+ * Helper for dealing with PHP's $_FILES. Instead of working with a sometimes puzzling array, with file info
+ * dispersed throughout, creates a single array of FileSubmissionInterface objects. This array can be multi-dimensional,
+ * but its shape follows that of $_POST and $_GET.
+ * Eg, access all file info for a file input named "file1" using
+ *
+ * ```
+ * $data_handler = LoaderFactory::getLoader()->load(' EventEspresso\core\services\request\files\FilesDataHandler');
+ * $files = $data_handler->getFileObjects();
+ * $file = $files['file1'];
+ * ```
+ *
+ * and for a file input named "my[great][file][input]", use the same code but change the last line to:
+ *
+ * ```
+ * $file = $files['my']['great']['file']['input'];
+ * ```
+ *
+ * In both cases, $file will be a FileSubmissionInterface object, containing the file's name, temporary filepath, size,
+ * error code, and helpers to get its mime type and extension.
+ *
+ *
+ *
+ * @package     Event Espresso
+ * @author         Mike Nelson
+ * @since         $VID:$
+ *
+ */
+class FilesDataHandler
+{
+    /**
+     * @var Request
+     */
+    protected $request;
+
+    /**
+     * @var array
+     */
+    protected $file_objects;
```

Comment:
:joy: I thought you said, in Slack DM, that you preferred a hierarchical array!

> mnelson4 [11:33 AM]
> hey brent, which do you think would be a better way to organize the files? in a hierarchical tree or a flat array (or collection)?...

> br3nt [11:56 AM]
> ya I think I would go with the hierarchical array too...

Would you rather a hierarchical collection instead then? That seems... unnatural :japanese_ogre: 

FYI I was planning on doing the validation on the files in the forms code. This is just a better way of organizing the files data (which, btw, was a major pain to get into a sane format).

Reply:
oh oops... ya... sry about that... didn't mean to filp flop...

I saw `isStrangeFilesArray()` and got it in my head that you had changed your mind and were flattening the arrays somewhat, but I see now that `organizeFilesData()` and `createFileObjects()` are actually recursive. I obviously read those wrong.

For what it's worth, you can use anything as an identifier for an object in a collection, so you _could_ keep the original forms data array "path" info intact by using that as the identifier while looping through form data (if that's how you want to be able to access the files), ex:

```php
$flattened_key = 'my[hierarchical][form]'; // built up recursively as you loop thru data
foreach ($form_data as $key => $value) {
	// recurse if value is another layer of form data, or:
	if ( isset( $value['name'], etc... ) ) {
		$file_submission = new FileSubmission( $value['name'], etc... );
		$files_collection->add( $file_submission, "{$flattened_key}[{$key}]" );
	}
}
```

which would allow you to find a specific `FileSubmission` object using a flattened representation of the form data structure:

```php
$files_collection->get( 'my[hierarchical][form][file_input]' );
```

but because the Collection is not a multidimensional array, you can ALSO loop through ALL of the entities it contains, if you simply want direct sequential access:

```php
foreach( $files_collection as $file_submission ) {
	// do something with file, like.. maybe find out where it was originally in the form data:
	$orig_path = $file_submission->getInfo(); // 'my[hierarchical][form][file_input]'
}
```

so using a Collection would kinda give you the best of both worlds, meaning direct sequential access, AND the ability to access items while looping through a multidimensional array. That's on top of the other benefits I mentioned as well, like restricting the collection to files of a particular type.

Reply:
kk no worries! I made the switch to using a 1d collection. That should work fine.